cmd/upgrade.go

```diff
@@ -0,0 +1,127 @@
+package cmd
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/spf13/cobra"
+)
+
+var upgradeCmd = &cobra.Command{
+	Use:   "upgrade",
+	Short: "Upgrade osdctl",
+	Long:  "Fetch latest osdctl from GitHub and replace the running binary",
+	RunE:  upgrade,
+}
+
+func upgrade(cmd *cobra.Command, args []string) error {
+	latest, err := getLatestVersion()
+	if err != nil {
+		return err
+	}
+	latestWithoutPrefix := strings.TrimPrefix(latest, "v")
+	// check if an upgrade is necessary
+	currentSemVer := semver.New(Version)
+	latestSemVer := semver.New(latestWithoutPrefix)
+	if currentSemVer.LessThan(*latestSemVer) {
+		// upgrade necessary
+		client := http.Client{
+			Timeout: time.Second * 60,
+		}
+
+		addr := fmt.Sprintf(versionAddressTemplate, latestWithoutPrefix, latestWithoutPrefix, parseGOOS(runtime.GOOS), parseGOARCH(runtime.GOARCH))
+
+		req, err := http.NewRequest(http.MethodGet, addr, nil)
+		if err != nil {
+			return err
+		}
+
+		res, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+
+		gzf, err := gzip.NewReader(res.Body)
+		if err != nil {
+			return err
+		}
+
+		tr := tar.NewReader(gzf)
+		for {
+			f, err := tr.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			if f.Name == "osdctl" {
+				homeDir, err := os.UserHomeDir()
+				if err != nil {
+					return err
+				}
+
+				dir, err := ioutil.TempDir(homeDir, ".*")
```

Comment:
Could you give an example to where this is created?

Reply:
It creates a temporary directory in the user's HOME directory. First, I went for a temporary directory in the ramfs (f.e. `/tmp`), but the problem was that it is impossible to replace the binary this way. 

For replacing a binary/executable on POSIX systems (MacOS/Linux/BSD) we have to do the move syscall.
The move syscall has one strict requirement: The old file location and the new file location must be on the same filesystem... that's why I went for the HOME directory instead.

But this also means: If you are running osdctl from a different filesystem, the whole operation may fail.

Long story, short: it's complicated :/

Reply:
One addition: Using a hard copy operation does not work. We have to copy the binary first to the target filesystem and then do a "move" or "rename" operation.

Reply:
@georgettica  I have added an explanation as comment in the code. Good catch! This definitely needed more explanation on why we are doing this, this way! :tada:

Reply:
@georgettica done